src/Controller/Component/AuthComponent.php

```diff
@@ -402,14 +402,21 @@ protected function _unauthenticated(Controller $controller)
      */
     protected function _loginActionRedirectUrl()
     {
-        $currentUrl = $this->request->here(false);
+        $urlToRedirectBackTo = $this->request->here(false);
+        if (!$this->request->is('get')) {
+            $urlToRedirectBackTo = $this->request->referer(true);
```

Comment:
This isn't guaranteed to be a GET. The user could be in a chain of POST redirects.